crates/messages/proto/messages.proto

```diff
@@ -31,137 +19,149 @@ message WorkerState {
+  string dataset = 2;  // "czM6Ly9ldGhlcmV1bS1tYWlubmV0"
+  string query = 3;  // 512 KB max
   // If present, these values should be used instead of from_block and to_block in the query contents
   optional Range block_range = 7;
+  uint64 timestamp_ms = 8;
+  bytes signature = 9;  // covers all the above fields + worker's peer ID
 }
 
+// Worker -> Portal
 message QueryResult {
   string query_id = 1;
   oneof result {
-    OkResult ok = 2;
-    string bad_request = 3;
-    string server_error = 4;
-    google.protobuf.Empty no_allocation = 5;
-    google.protobuf.Empty timeout_v1 = 6 [deprecated = true];
-    string timeout = 7;
+    QueryOk ok = 2;
+    QueryError err = 8;
   }
+  optional uint32 retry_after_ms = 9;
 }
 
-message OkResult {
+message QueryOk {
   bytes data = 1;
-  optional bytes exec_plan = 2;
   optional uint64 last_block = 3;
+  QueryResultSummary summary = 4;
 }
 
-message QuerySubmitted {
-  string client_id = 1;
-  string worker_id = 2;
-  string query_id = 3;
-
-  string dataset = 4;
-  string query = 5;
-  bytes query_hash = 6;
+message QueryError {
+  oneof err {
+    string bad_request = 1;
+    string not_found = 2;
+    string server_error = 3;
+    google.protobuf.Empty too_many_requests = 4;
+    google.protobuf.Empty server_overloaded = 5;
+    string timeout = 6;
+  }
 }
 
+// Portal -> Logs collector
 message QueryFinished {
-  string client_id = 1;
   string worker_id = 2;
   string query_id = 3;
 
-  uint32 exec_time_ms = 4;
+  uint32 total_time_ms = 4;
   oneof result {
-    SizeAndHash ok = 5;
-    string bad_request = 6;
-    string server_error = 7;
-    google.protobuf.Empty timeout_v1 = 8  [deprecated = true];
-    google.protobuf.Empty no_allocation = 9;
-    string timeout = 10;
+    QueryResultSummary ok = 5;
+    QueryError err = 11;
   }
 }
 
-message QueryExecuted {
-  string client_id = 1;
-  string worker_id = 2;
-
-  Query query = 3;
-  bytes query_hash = 5;
-
-  optional uint32 exec_time_ms = 6; // optional to enforce serializing 0
-  oneof result {
-    InputAndOutput ok = 7;
-    string bad_request = 8;
-    string server_error = 9;
-  }
-  optional uint64 seq_no = 10; // optional to enforce serializing 0
-  optional uint64 timestamp_ms = 11; // optional to enforce serializing 0
-  bytes signature = 12;
+// Logs collector -> Worker
+message LogsRequest {
+  uint64 from_timestamp_ms = 1;
+  uint64 to_timestamp_ms = 2;
+  optional string last_received_query_id = 3;  // query ID of the last collected query
 }
 
+// Worker -> Logs collector
 message QueryLogs {
   repeated QueryExecuted queries_executed = 1;
+  bool has_more = 2;
 }
 
-message InputAndOutput {
-  optional uint32 num_read_chunks = 1; // optional to enforce serializing 0
-  SizeAndHash output = 2;
-}
+message QueryExecuted {
+  string client_id = 1;
+  Query query = 3;
 
-message SizeAndHash {
-  optional uint32 size = 1; // optional to enforce serializing 0
-  bytes sha3_256 = 2;
+  uint32 exec_time_ms = 6;
```

Comment:
No hope to complete under 1 ms? :)

On a serious note, the precision is a bit low, because we have meaningful queries that complete in just few `ms`

What those times will measure exactly?

Reply:
Okay, I'll change it, it just was like this forever.\
I think it's purely statistical — to have a better understanding of the network performance.